homeassistant/helpers/script_variables.py

```diff
@@ -65,6 +60,216 @@ def async_render(
+        )
+
+    def exit_scope(self) -> ScriptRunVariables:
+        """Exit the current scope.
+
+        Does no clean-up, but simply returns the previous scope.
+        """
+        if self._previous is None:
+            raise ValueError("Cannot exit root scope")
+        return self._previous
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        """Assign value to a variable. Equivalent to `assign_single`."""
+        self.assign_single(key, value)
+
+    def __delitem__(self, key: str) -> None:
+        """Delete a variable (disallowed)."""
+        raise TypeError("Deleting items is not allowed in ScriptRunVariables.")
+
+    def assign_single(
+        self, key: str, value: Any, *, parallel_special: bool = False
+    ) -> None:
+        """Assign value to a variable.
+
+        Value is always assigned to the variable in the nearest scope, in which it is defined.
+        If the variable is not defined at all, it is created in the top-level scope.
+
+        :param parallel_special: Whether variable is to be treated specially in parallel sequences.
+        """
+        if self._local_store is not None and key in self._local_store:
+            self._local_store[key] = value
+            return
+
+        if self._parent is None:
+            assert self._local_store is not None  # top level always has a local store
+            self._local_store[key] = value
+            return
+
+        if self._parallel_store is not None:
+            self._parallel_store.outer_scope_writes[key] = value
+            if parallel_special:
+                self._parallel_store.special[key] = value
+            else:
+                self._parallel_store.special.pop(key, None)
+
+        self._parent.assign_single(key, value, parallel_special=parallel_special)
+
+    def assign(self, new_vars: ScriptVariables) -> None:
+        """Assign rendered values to variables.
+
+        First renders the variable values in order and then assigns each of them.
+        Value is always assigned to the variable in the nearest scope, in which it is defined.
+        If the variable is not defined at all, it is created in the top-level scope.
+        """
+        for key, value in new_vars.async_simple_render(self).items():
+            self.assign_single(key, value)
+
+    def define_single(self, key: str, value: Any) -> None:
+        """Define a local variable and assign value to it."""
+        self._ensure_local()
+        assert self._local_store is not None
+        self._local_store[key] = value
+
+    def define(self, new_vars: ScriptVariables) -> None:
+        """Define local variables and assign rendered values to them.
+
+        First renders the variable values in order and then assigns each of them.
+        """
+        self._ensure_local()
+        assert self._local_store is not None
+        for key, value in new_vars.async_simple_render(self).items():
+            self._local_store[key] = value
```

Comment:
Why does our special dict class need to know about Jinja rendering, it's called only a single time from the script helper?